File: androidApp/src/main/java/com/kukhtoslava/weatherapp/android/ui/description/DescriptionScreen.kt
package com.kukhtoslava.weatherapp.android.ui.description

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.ExperimentalLifecycleComposeApi
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.navigation.NavHostController
import com.kukhtoslava.weatherapp.android.R
import com.kukhtoslava.weatherapp.android.coreui.*
import com.kukhtoslava.weatherapp.android.uihelpers.WeatherPartItem
import com.kukhtoslava.weatherapp.android.utils.convertIconsToResources
import com.kukhtoslava.weatherapp.android.utils.convertSecondsToDaysString
import com.kukhtoslava.weatherapp.android.utils.convertSecondsToHoursMinutesString
import com.kukhtoslava.weatherapp.android.utils.rememberFlowWithLifecycle
import com.kukhtoslava.weatherapp.presentation.description.DescriptionAction
import com.kukhtoslava.weatherapp.presentation.description.DescriptionEvent
import com.kukhtoslava.weatherapp.presentation.description.DescriptionState
import com.kukhtoslava.weatherapp.presentation.description.DescriptionViewModel

@OptIn(ExperimentalLifecycleComposeApi::class)
@Composable
fun DescriptionScreen(
    navHostController: NavHostController,
    viewModel: DescriptionViewModel
) {

    val state by viewModel.stateFlow.collectAsStateWithLifecycle()
    val eventFlow = rememberFlowWithLifecycle(flow = viewModel.eventFlow)

    BackHandler(enabled = true) {
        navHostController.popBackStack()
    }
    LaunchedEffect(Unit) {
        viewModel.dispatch(DescriptionAction.Load)
    }
    LaunchedEffect(eventFlow) {
        eventFlow.collect { event ->
            when (event) {
                DescriptionEvent.Close -> navHostController.popBackStack()
            }
        }
    }

    DescriptionUI(
        state = state,
        closeClicked = { viewModel.dispatch(action = DescriptionAction.Close) }
    )
}

@Composable
fun DescriptionUI(
    state: DescriptionState,
    closeClicked: () -> Unit
) {

    val context = LocalContext.current

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(
                brush = mainBrush
            )
            .statusBarsPadding()
    ) {

        val place = state.placeName
        val temp = state.fullWeather?.current?.temp?.toInt() ?: "--"
        val description = state.fullWeather?.current?.weather?.get(0)?.description ?: ""
        val fullDescription = context.getString(
            R.string.description_title,
            temp,
            description
        )
        val sunrise = state.fullWeather?.current?.sunrise?.let { dt ->
            convertSecondsToHoursMinutesString(dt)
        } ?: "--"
        val sunset = state.fullWeather?.current?.sunset?.let { dt ->
            convertSecondsToHoursMinutesString(dt)
        } ?: "--"
        val feelLikes = state.fullWeather?.current?.feelsLike?.toInt()?.let { temperature ->
            context.getString(
                R.string.degree,
                temperature.toString()
            )
        } ?: "--"
        val uviIndex = state.fullWeather?.current?.uvi?.toString() ?: "--"
        val humidity = state.fullWeather?.current?.humidity?.let { hum ->
            context.getString(
                R.string.humidity_percent,
                hum.toString()
            )
        } ?: "--"
        val wind = state.fullWeather?.current?.windSpeed?.let { wind ->
            context.getString(
                R.string.speed,
                wind.toString()
            )
        } ?: "--"
        val visibility = state.fullWeather?.current?.visibility?.let { vis ->
            val visInKm = vis / 1000
            context.getString(
                R.string.visibility_meters,
                visInKm.toString()
            )
        } ?: "--"
        val pressure = state.fullWeather?.current?.pressure?.let { press ->
            context.getString(
                R.string.pressure_metric,
                press.toString()
            )
        } ?: "--"


        Box(
            modifier = Modifier
                .fillMaxWidth(),
            contentAlignment = Alignment.CenterEnd
        ) {
            Text(
                modifier = Modifier
                    .clickable { closeClicked() }
                    .padding(16.dp),
                text = stringResource(id = R.string.close),
                style = Type.body3,
                maxLines = 1
            )
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .statusBarsPadding()
        ) {

            Text(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = 8.dp, end = 8.dp, top = 8.dp),
                text = place,
                style = Type.header2,
                maxLines = 1
            )

            Text(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 8.dp),
                text = fullDescription,
                style = Type.header4,
                maxLines = 1
            )

            Column(
                modifier = Modifier
                    .padding(start = 8.dp, end = 8.dp, bottom = 8.dp)
                    .fillMaxWidth()
                    .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                    .clip(WeatherShapes.extraLarge)
                    .background(color = PortGore)
            ) {

                Text(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(22.dp),
                    text = stringResource(id = R.string.weekly_forecast),
                    style = Type.header5,
                    maxLines = 1
                )

                Row(
                    modifier = Modifier
                        .padding(all = 8.dp)
                        .horizontalScroll(rememberScrollState())
                        .fillMaxWidth()
                ) {
                    state.fullWeather?.daily?.forEach { item ->
                        val temperature = context.getString(
                            R.string.degree,
                            (item.temp.day.toInt()).toString()
                        )
                        WeatherPartItem(
                            header = convertSecondsToDaysString(item.dt).uppercase(),
                            id = convertIconsToResources(item.weather[0].icon),
                            temperature = temperature
                        )
                    }
                }
            }

            Row(
                modifier = Modifier
                    .padding(bottom = 8.dp)
                    .fillMaxWidth()
            ) {

                Column(
                    modifier = Modifier
                        .padding(start = 8.dp, end = 4.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.sunrise),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = sunrise,
                        style = Type.body7
                    )
                }

                Column(
                    modifier = Modifier
                        .padding(start = 4.dp, end = 8.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.sunset),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        style = Type.body7,
                        text = sunset
                    )
                }
            }

            Row(
                modifier = Modifier
                    .padding(bottom = 8.dp)
                    .fillMaxWidth()
            ) {

                Column(
                    modifier = Modifier
                        .padding(start = 8.dp, end = 4.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.feels_like),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = feelLikes,
                        style = Type.body7
                    )
                }

                Column(
                    modifier = Modifier
                        .padding(start = 4.dp, end = 8.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.uvi),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = uviIndex,
                        style = Type.body7
                    )
                }
            }

            Row(
                modifier = Modifier
                    .padding(bottom = 8.dp)
                    .fillMaxWidth()
            ) {

                Column(
                    modifier = Modifier
                        .padding(start = 8.dp, end = 4.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.wind),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = wind,
                        style = Type.body7
                    )
                }

                Column(
                    modifier = Modifier
                        .padding(start = 4.dp, end = 8.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.humidity),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = humidity,
                        style = Type.body7
                    )
                }
            }

            Row(
                modifier = Modifier
                    .padding(bottom = 8.dp)
                    .fillMaxWidth()
            ) {

                Column(
                    modifier = Modifier
                        .padding(start = 8.dp, end = 4.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.visibility),
                        color = Whisper50,
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = visibility,
                        color = Color.White,
                        style = Type.body7
                    )
                }

                Column(
                    modifier = Modifier
                        .padding(start = 4.dp, end = 8.dp)
                        .weight(1f)
                        .aspectRatio(1f)
                        .border(2.dp, color = Gigas, shape = WeatherShapes.extraLarge)
                        .clip(WeatherShapes.extraLarge)
                        .background(color = PortGore)
                ) {
                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(22.dp),
                        text = stringResource(id = R.string.pressure),
                        style = Type.header5,
                        maxLines = 1
                    )

                    Text(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(start = 22.dp),
                        text = pressure,
                        style = Type.body7
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/ui/description/DescriptionScreen.swift
//
//  DescriptionScreen.swift
//  iosApp
//
//  Created by Slava Kukhto on 05/10/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import mokoMvvmFlowSwiftUI
import MultiPlatformLibrary

struct DescriptionScreen: View {
    
    @EnvironmentObject var globalDataHolder: GlobalDataHolder
    
    @ObservedObject var viewModel: DescriptionViewModel = DescriptionViewModel()
    @State private var state: DescriptionState = DescriptionState(placeName: "", fullWeather: nil)
    
    var body: some View {
        VStack {
            HStack {
                Spacer()
                Button(action: {
                    viewModel.dispatch(action: DescriptionActionClose())
                }) {
                    Text("close")
                        .font(.system(size: 15))
                        .fontWeight(.semibold)
                        .foregroundColor(.white)
                        .padding(16)
                }
            }
            prepareDescriptionContent(state: self.state)
        }
        .edgesIgnoringSafeArea(.bottom)
        .background(
            LinearGradient(gradient: Gradient(colors: ColorUtils.MainBrushColors), startPoint: .top, endPoint: .bottom)
        )
        .onReceive(createPublisher(viewModel.stateFlow)) { state in
            self.state = state
        }
        .onReceive(createPublisher(viewModel.eventFlow)) { event in
            let eventKt = DescriptionEventKs(event)
            switch(eventKt) {
            case .close:
                withAnimation {
                    globalDataHolder.currentAppScreen = .main
                }
                break
            }
        }
        .onAppear {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3, execute: {
                viewModel.dispatch(action: DescriptionActionLoad())
            })
        }
    }
}

extension DescriptionScreen {
    
    func prepareDescriptionContent(state: DescriptionState) -> some View {
        
        let place = state.placeName
        let fullDescription: String
        let sunrise: String
        let sunset: String
        let feelLikes: String
        let uviIndex: String
        let humidity: String
        let wind: String
        let visibility: String
        let pressure: String
        let daily: [Daily]
        
        if (state.fullWeather != nil) {
            let temp = (state.fullWeather?.current.temp.toInt())
            let description = state.fullWeather?.current.weather[0].description_
            fullDescription = String(format: NSLocalizedString("description_title", comment: "Full description"), temp!, description!)
            
            sunrise = convertSecondsToHoursMinutesString(dt: Int(state.fullWeather!.current.sunrise))
            sunset = convertSecondsToHoursMinutesString(dt: Int(state.fullWeather!.current.sunset))
            
            let feelLikesTemperature = state.fullWeather?.current.feelsLike.toInt()
            feelLikes = String(format: NSLocalizedString("degree", comment: "Degree"), feelLikesTemperature!)
            
            uviIndex = String("\(state.fullWeather!.current.uvi)")
            
            humidity = String(format: NSLocalizedString("humidity_percent", comment: "Humidity percent"), state.fullWeather!.current.humidity)
            
            wind = String(format: NSLocalizedString("speed", comment: "Wind speed"), String("\(state.fullWeather!.current.windSpeed)"))
            
            visibility = String(format: NSLocalizedString("visibility_meters", comment: "Visibility"), String("\(((state.fullWeather!.current.visibility)) / 1000)"))
            
            pressure = String(format: NSLocalizedString("pressure_metric", comment: "Pressure"), String("\(state.fullWeather!.current.pressure)"))
            
            daily = state.fullWeather?.daily ?? []
            
        } else {
            fullDescription = ""
            sunrise = "--"
            sunset = "--"
            feelLikes = "--"
            uviIndex = "--"
            humidity = "--"
            wind = "--"
            visibility = "--"
            pressure = "--"
            daily = []
        }
        
        return ScrollView {
            VStack {
                Text(place)
                    .font(.system(size: 34))
                    .fontWeight(.medium)
                    .foregroundColor(.white)
                    .padding([.leading, .top, .trailing], 8)
                Text(fullDescription)
                    .font(.system(size: 18))
                    .fontWeight(.semibold)
                    .foregroundColor(Color(ColorUtils.Whisper50))
                    .padding([.leading, .trailing], 8)
                VStack {
                    HStack {
                        Text("weekly_forecast")
                            .font(.system(size: 15))
                            .fontWeight(.semibold)
                            .foregroundColor(Color(ColorUtils.Whisper50))
                            .padding([.leading, .top, .trailing], 20)
                        Spacer()
                    }
                    HStack {
                        ScrollView(.horizontal) {
                            LazyHStack {
                                ForEach(daily, id: \.self) { day in
                                    let header = convertSecondsToDaysString(dt: Int(day.dt)).uppercased()
                                    let resource = convertIconsToResources(icon: day.weather[0].icon)
                                    let temp = String(format: NSLocalizedString("degree", comment: "Degree"), day.temp.day.toInt() ?? "--")
                                    WeatherPartItem(header: header, resourceName: resource, temperature: temp)
                                        .padding(.leading, 4)
                                        .padding(.trailing, 4)
                                }
                            }
                        }
                    }
                    .padding(20)
                }
                .frame(height: 250)
                .background(Color(ColorUtils.PortGore))
                .cornerRadius(28)
                .overlay(RoundedRectangle(cornerRadius: 28)
                    .stroke(Color(ColorUtils.Gigas), lineWidth: 2.0))
                
                HStack {
                    prepareItemView(title: "sunrise", value: sunrise)
                    prepareItemView(title: "sunset", value: sunset)
                }
                
                HStack {
                    prepareItemView(title: "feels_like", value: feelLikes)
                    prepareItemView(title: "uvi", value: uviIndex)
                }
                
                HStack {
                    prepareItemView(title: "wind", value: wind)
                    prepareItemView(title: "humidity", value: humidity)
                }
                
                HStack {
                    prepareItemView(title: "visibility", value: visibility)
                    prepareItemView(title: "pressure", value: pressure)
                }
            }
            .padding(16)
        }
    }
    
    func prepareItemView(title: String, value: String) -> some View {
        
        let width = (UIScreen.main.bounds.width - 32) / 2
        
        return VStack {
            HStack {
                Text(title.localized)
                    .font(.system(size: 15))
                    .fontWeight(.semibold)
                    .foregroundColor(Color(ColorUtils.Whisper50))
                    .padding([.leading, .top, .trailing], 20)
                Spacer()
            }
            HStack {
                Text(value)
                    .font(.system(size: 32))
                    .fontWeight(.semibold)
                    .foregroundColor(.white)
                    .padding([.leading, .top, .trailing], 16)
                Spacer()
            }
            Spacer()
        }
        .frame(width: width, height: width)
        .background(Color(ColorUtils.PortGore))
        .cornerRadius(28)
        .overlay(RoundedRectangle(cornerRadius: 28)
            .stroke(Color(ColorUtils.Gigas), lineWidth: 2.0))
    }
}
